### PR TITLE
Drainage regen simplification and other fixes

### PR DESCRIPTION
**Pull request by Zematus**
This PR is mostly focused on fixing bug #200 though I ended up disabling automatic drainage regeneration for brushes since I couldn't make it work fine after several attempts. Instead. I ended up adding a manual regeneration button that does a full world regeneration. It guarantees that the end result doesn't have invalid drainage values and all values are identical even after a reload. 

I also ended up simplifying regeneration processes by making all of them go through the same steps. And now, loaded worlds will regenerate most cells values regardless if the cell was modified or not, simplifying the save file a little bit.

This PR also includes a partial fix for issue #197. It is not perfect because the exported image does not contain shader data. Making it look simpler and uglier than what's seen within the game.

### CODE REVIEW

**Comment by Zematus** [comment]
Sorry for the long PR. It took me a while to get things to work and I left a little bit of dead code. I also tried to add some documentation but I had no time to go through all the changes.

Take your time reviewing. Don't rush it, but also don't sweat on the details.

File: Assets/Scripts/2D/GuiManagerScript.cs [hunk 1]
```diff
@@ -247,7 +251,7 @@ public void DeregisterLoadWorldPostProgressOp(PostProgressOperation op)
         _loadWorldPostProgressOp -= op;
     }
 
-    public void ResetAllDialogs()
+    private void ResetAllDialogs()
```
Comment (nsmonagh) [inline Assets/Scripts/2D/GuiManagerScript.cs]: Why were these set to private?

File: Assets/Scripts/2D/GuiManagerScript.cs [hunk 2]
```diff
@@ -2005,8 +2017,8 @@ public void PostProgressOp_LoadAction()
 
         _postProgressOp -= PostProgressOp_LoadAction;
 
-        if (_loadWorldPostProgressOp != null)
-            _loadWorldPostProgressOp.Invoke();
+        _loadWorldPostProgressOp?.Invoke();
```
Comment (nsmonagh) [inline Assets/Scripts/2D/GuiManagerScript.cs]: I like all the null propagation

File: Assets/Scripts/2D/GuiManagerScript.cs [hunk 3]
```diff
@@ -1304,33 +1312,33 @@ public void RegenerateWorldRiverLevelOffsetChange(float value)
     public void RegenerateWorldTemperatureOffsetChange(float value)
     {
         Manager.TemperatureOffset = value;
-
-        RegenerateWorld(GenerationType.TemperatureRegeneration);
```
Comment (nsmonagh) [inline Assets/Scripts/2D/GuiManagerScript.cs]: Everything appropriately changed to TerrainRegeneration since whole world has to be regenerated.

File: Assets/Scripts/2D/MapEditor/MapEditorToolbarScript.cs [hunk 4]
```diff
@@ -367,6 +370,15 @@ public void RegenerateWorldLayerNoiseInfluenceChange(float value)
     private void OnUndoStackUpdate()
     {
         UndoActionButton.interactable = Manager.UndoableEditorActionsCount > 0;
+
+        if (Manager.CurrentWorld.NeedsDrainageRegeneration)
```
Comment (nsmonagh) [inline Assets/Scripts/2D/MapEditor/MapEditorToolbarScript.cs]: Drainage regeneration check looks good.

File: Assets/Scripts/2D/MapScript.cs [hunk 5]
```diff
@@ -210,6 +212,7 @@ public void PointerDown(BaseEventData data)
             if (!Manager.ViewingGlobe && !_isDraggingMap)
             {
                 Manager.ActivateEditorBrush(true);
+                ActivatedBrush.Invoke(true);
```
Comment (nsmonagh) [inline Assets/Scripts/2D/MapScript.cs]: What does this do?

File: Assets/Scripts/2D/ModalPanels/AddPopulationDialogScript.cs [hunk 6]
```diff
@@ -17,15 +17,16 @@ void Update()
         ReadKeyboardInput();
     }
 
+    /// <summary>Handler for changes to the population input field.</summary>
```
Comment (nsmonagh) [inline Assets/Scripts/2D/ModalPanels/AddPopulationDialogScript.cs]: Thank you!

File: Assets/Scripts/WorldEngine/Manager.cs [hunk 7]
```diff
@@ -926,49 +924,65 @@ public static Vector2 GetUVFromMapCoordinates(WorldPosition mapPosition)
...
 
             for (int i = 0; i < width; i++)
             {
                 for (int j = 0; j < height; j++)
                 {
-
                     int finalX = (i + xOffset) % width;
 
-                    exportTexture.SetPixel(i, j, mapTexture.GetPixel(finalX, j));
+                    Color mapPixelColor = mapTexture.GetPixel(finalX, j);
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/Manager.cs]: This is a good addition

File: Assets/Scripts/WorldEngine/Terrain/TerrainCell.cs [hunk 8]
```diff
@@ -91,7 +91,8 @@ public class TerrainCell
     public float Buffer3 = 0;
     public int RiverId = -1;
     public float RiverLength = 0;
-    public bool DrainageDone = false;
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/Terrain/TerrainCell.cs]: Why did this change to true?

File: Assets/Scripts/WorldEngine/World.cs [hunk 9]
```diff
@@ -487,9 +489,11 @@ public World(int width, int height, int seed)
         TerrainCellAlterationListCount = 0;
     }
 
-    private void InitializeTerrainLimitsAndSettings()
+    /// <summary>Initializes the parameters to be used when generating a new world.</summary>
+    /// <param name="justLoaded">Indicate if the world being generated is based on a loaded save file.</param>
+    private void InitializeTerrainLimitsAndSettings(bool justLoaded)
     {
-        _justLoaded = false;
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/World.cs]: Good change

File: Assets/Scripts/WorldEngine/Manager.cs [hunk 10]
```diff
@@ -983,9 +997,22 @@ public static void ExportMapTextureToFileAsync(string path, Rect uvRect, Progres
...
+            }
+            catch (System.Exception e)
+            {
+                // To display the exception on screen we need to queue a task on the rendering engine's main thread
+                // and rethrow the exception inside that task
+                EnqueueTaskAndWait(() =>
+                {
+                    throw new System.Exception("Unhandled exception in ExportMapTextureToFile with path: " + path, e);
+                });
+            }
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/Manager.cs]: Good addition to error handling.

File: Assets/Scripts/WorldEngine/Terrain/TerrainCell.cs [hunk 11]
```diff
@@ -287,21 +288,6 @@ public void SetAlteration(TerrainCellAlteration alteration)
...
-        if (!alteration.Temperature.IsInsideRange(World.MinPossibleTemperatureWithOffset - 0.5f, World.MaxPossibleTemperatureWithOffset + 0.5f))
-        {
-            Debug.Log("Invalid alteration.Temperature: " + alteration.Temperature);
-        }
-#endif
-
-        Temperature = alteration.Temperature;
-        OriginalTemperature = alteration.OriginalTemperature;
-        Rainfall = alteration.Rainfall;
-        WaterAccumulation = alteration.WaterAccumulation;
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/Terrain/TerrainCell.cs]: Why were these variables removed? Not used?

File: Assets/Scripts/WorldEngine/World.cs [hunk 12]
```diff
@@ -2193,9 +2197,14 @@ public void GenerateTerrain(GenerationType type, Texture2D heightmap)
...
         ProgressCastMethod(_accumulatedProgress, "Generating Drainage Basins...");
 
+        SetCellsToEvalForDrainange();
+
         GenerateDrainageBasins();
         GenerateDrainageBasins(false); // repeat to simulate geological scale erosion
-        
+
+        _cellsToDrain.Clear();
+        NeedsDrainageRegeneration = false;
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/World.cs]: What does the addition of these lines do?

File: Assets/Scripts/WorldEngine/World.cs [hunk 13]
```diff
@@ -2560,43 +2569,34 @@ public void ModifyCellAltitude(TerrainCell cell, float valueOffset, float noiseF
...
+            TerrainCell cell = _drainageEvalHeap.Extract(false);
 
             if (!cell.DrainageDone)
             {
                 bool reInsert = false;
 
-                UpdateWater(cell);
-
-                float cellAltitude = Mathf.Max(0, CalculateAltitudePlusAccumulation(cell));
+                float cellAltitude = Mathf.Max(0, cell.Altitude);
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/World.cs]: Why the removal of accumulation?

File: Assets/Scripts/WorldEngine/World.cs [hunk 14]
```diff
@@ -2606,41 +2606,20 @@ public void PerformTerrainAlterationDrainageRegen(bool doTerrainModification = t
...
-                {
-                    cellsToModify.Add(cell);
-                }
+                UpdateWater(cell);
             }
 
             DrainToNeighbors(cell, AddToDrainageRegen);
         }
-
-        if (doTerrainModification)
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/World.cs]: Also why the removal of terrain modification?

sidenote: I know you said not to sweat the details, but the drainage feature is complex and I'm just trying to understand.

File: Assets/Scripts/WorldEngine/World.cs [hunk 15]
```diff
@@ -3323,151 +3313,6 @@ private void GenerateTerrainRainfall(TerrainCell cell, bool justSetRainfallSourc
         CalculateAndSetRainfall(cell, rainfallValue);
     }
 
-    private void GenerateTerrainRainfall2(TerrainCell cell, bool justSetRainfallSources = false, bool setDependencies = true)
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/World.cs]: Why was this removed?

File: Assets/Scripts/WorldEngine/World.cs [hunk 16]
```diff
@@ -3673,94 +3523,103 @@ private void DrainToNeighbors(TerrainCell cell, addToCellsToDrainDelegate addToC
...
         BinaryHeap<TerrainCell> cellsToDrain = new BinaryHeap<TerrainCell>(TerrainCell.CompareOriginalAltitude, 500000);
 
-        //Debug.Log("Generating Rivers - River Strength: " + RiverStrength);
+        int i = 0;
+        int evalCount = _cellsToDrain.Count;
+        int segSize = 100;
 
         // find cells that are higher than neighbors
-        for (int i = 0; i < sizeX; i++)
+        foreach (TerrainCell cell in _cellsToDrain)
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/World.cs]: Nice

**Comment by nsmonagh** [comment]
Overall I feel like I at least half understood what was going on heh

There were some changes that I was puzzled by, mostly the removal of that large function.

All the documentation is very useful and it makes me want to go back and rewrite some of mine since I feel mine is too generic and too influenced by the function title. So I just need to get better.

After the bugs with drainage systems are fixed will world regeneration revert to only regenerating what is needed?

Anyways, bug fixes in general are a pain so I appreciate the solutions you have made.

File: Assets/Scripts/2D/GuiManagerScript.cs (hunk 1 above)
Comment (Zematus) [reply Assets/Scripts/2D/GuiManagerScript.cs]: Those function were not intended to be used outside of the class. Beside that, if they are public they become visible to the Unity3D editor, which is not really good when the class way too many public functions. Unity3D is not designed as IDE and it sucks at discoverability. It becomes a chore trying to find specific functions in unity if there are too many of them exposed.

File: Assets/Scripts/2D/GuiManagerScript.cs (hunk 2 above)
Comment (Zematus) [reply Assets/Scripts/2D/GuiManagerScript.cs]: Thanks, I saw you using it and it's way better than adding if(null) blocks

File: Assets/Scripts/2D/MapScript.cs (hunk 5 above)
Comment (Zematus) [reply Assets/Scripts/2D/MapScript.cs]: ActivatedBrush is an Unity event that is visible within the Unity3D editor. it's useful if you want to link events to listeners within the editor without having to add extra lines of code. There's one big donwside in the fact that this hides dependencies from the IDE/compiler.

In this case I'm using this event to control the states of the 'Drainage Regeneration', 'Exit Editor Mode' and 'Save' buttons in the UI

File: Assets/Scripts/WorldEngine/Manager.cs (hunk 10 above)
Comment (Zematus) [reply Assets/Scripts/WorldEngine/Manager.cs]: Yeah, I did the same for the other async tasks a while ago but I missed this particular case. Now it's taken care of.

File: Assets/Scripts/WorldEngine/Terrain/TerrainCell.cs (hunk 8 above)
Comment (Zematus) [reply Assets/Scripts/WorldEngine/Terrain/TerrainCell.cs]: Because for most cells there's no need to do drainage calculations if it receives no rainfall. If a cell receives rainfall or drainage from other cells then the value will be changed to false as needed.

File: Assets/Scripts/WorldEngine/Terrain/TerrainCell.cs (hunk 11 above)
Comment (Zematus) [reply Assets/Scripts/WorldEngine/Terrain/TerrainCell.cs]: They were used before. Originally, when a cell was edited in the editor, the changes would be saved to file in a "TileCellAlteration" object. But I found out that these changes where not being calculated correctly when using brushes and it was better to regenerate them globally once the brush became inactive, or after a file load. The only real values that need to be stored in file if changed are the ones prefixed with "Base..." The others can be derived from these.

File: Assets/Scripts/WorldEngine/World.cs (hunk 12 above)
Comment (Zematus) [reply Assets/Scripts/WorldEngine/World.cs]: The NeedsDrainageRegeneration flag is used to indicate to the editor if the user needs to manually regenerate the world before it can be saved or simulated with. A few operations alter this flag, Brush operations (which always enable the flag), Undos/Redos (which might or might not enable the flag) and Regenerations (which always disable the flag).

The _cellsToDrain hashset is used within the GenerateDrainageBasins to figure out which cells need drainage calculations/recalculations

File: Assets/Scripts/WorldEngine/World.cs (hunk 13 above)
Comment (Zematus) [reply Assets/Scripts/WorldEngine/World.cs]: That function was not very good. It's intention was to simulate water 'overflowing' from a cell. But the effect was not really noticeable and it introduced a host of issues in the over drainage calculation logic. The drainage calculation stack is supposed to have cells partially ordered from highest to lowest, but that function messed up with that order, forcing the need to reintroduce cells into the stack, which didn't always work as intended.

File: Assets/Scripts/WorldEngine/World.cs (hunk 14 above)
Comment (Zematus) [reply Assets/Scripts/WorldEngine/World.cs]: The intention here was to alter a cell's altitude or temperature due to drainage effects. But these alteration should not be done while drainage was still being processed for other cells. Instead, those alterations are now done after all drainage calculations have been completed.

File: Assets/Scripts/WorldEngine/World.cs (hunk 15 above)
Comment (Zematus) [reply Assets/Scripts/WorldEngine/World.cs]: This was an experimental function that was supposed to replace GenerateTerrainRainfall. But it was very slow and I didn't like the results. I kept the function because I wanted to rework it and make it better. But now the plan is to eventually introduce wind patterns, which will force a rewrite of rainfall generation in a completely different way. Therefore I don't want to keep a function that will never be used.

**Comment by Zematus** [comment]
Don't worry about the big functions. I have the bad habit of leaving unused code instead of removing it when it's not longer needed because I wonder that maybe I'll use it in the future (which is rarely the case). That big rainfall function has been there for perhaps a year and it was never used.

Honestly I don't think I will revert the world regeneration process to the way it was done before. The speed gains from skipping parts of the regeneration process where minimal and introduced many changes for bugs to happen (which they did). It's better to have a single monotonic process with little that always does the same, even if redundant, than a very parametric process with many confusing execution paths which might conflict with each other.
If anything, the terrain generation/regeneration function needs a throughout cleanup to remove all parameters and codepaths which are no longer going to be used. They just add noise to the code.

**Comment by Zematus** [comment]
Oh, by the way, thanks for the review!
